oonib/oonibackend.py

```diff
@@ -102,60 +102,71 @@
     multiService.addService(http_return_request_helper)
     http_return_request_helper.startService()
 
-# add the tor collector service here
+def getHSEndpoint(endpoint_config):
+    hsdir = os.path.join(torconfig.DataDirectory, endpoint_config['hsdir'])
+    if LooseVersion(txtorcon_version) >= LooseVersion('0.10.0'):
+        return TCPHiddenServiceEndpoint.global_tor(reactor,
+                                        80,
+                                        hidden_service_dir=hsdir)
+    else:
+        return TCPHiddenServiceEndpoint.global_tor(reactor,
+                                        80,
+                                        data_dir=hsdir)
+
+def getTCPEndpoint(endpoint_config):
+    return endpoints.TCP4ServerEndpoint(reactor, endpoint_config['port'])
+
+def getTLSEndpoint(endpoint_config):
+    with open(endpoint_config['cert'], 'r') as f:
+        cert_data = f.read()
+    certificate = ssl.PrivateCertificate.loadPEM(cert_data)
+    print certificate.inspect()
+    return endpoints.SSL4ServerEndpoint(reactor,
+                                        endpoint_config['port'],
+                                        certificate.options())
+
+def getEndpoint(endpoint_config):
+    if endpoint_config['type'] == 'onion':
+        return getHSEndpoint(endpoint_config)
+    elif endpoint_config['type'] == 'tcp':
+        return getTCPEndpoint(endpoint_config)
+    elif endpoint_config['type'] == 'tls':
+        return getTLSEndpoint(endpoint_config)
+    else:
+        raise Exception("unknown endpoint type")
+
+class Echo(protocol.Protocol):
+    def dataReceived(self, data):
+        self.transport.write('server echoes ' + data)
+
+def createService(endpoint, role, endpoint_config):
+    if role == 'bouncer':
+        factory = ooniBouncer
+    elif role == 'collector':
+        factory = ooniBackend
+    else:
+        raise Exception("unknown service type")
+
+    service = internet.StreamServerEndpointService(
+        endpoint, factory
+    )
+    service.setName("-".join([endpoint_config['type'], role]))
+    multiService.addService(service)
+    service.startService()
+
 if config.main.tor_hidden_service:
     torconfig = TorConfig()
-    d = startTor(torconfig)
-
-    def getHSEndpoint(data_dir):
-        if LooseVersion(txtorcon_version) >= LooseVersion('0.10.0'):
-            return TCPHiddenServiceEndpoint(reactor,
-                                            torconfig,
-                                            80,
-                                            hidden_service_dir=data_dir)
-        else:
-            return TCPHiddenServiceEndpoint(reactor,
-                                            torconfig,
-                                            80,
-                                            data_dir=data_dir)
-
-    def printOnionEndpoint(endpointService):
-        print ("Exposed %s Tor hidden service on httpo://%s" %
```

Comment:
We should maintain the ability to be able to see on which address the HS is exposed on.